### ISSUE TITLE
[FEATURE] Make 2 smaller screens merge to 1 large when opening larger screen on foldable devices

### ISSUE THREAD

**Issue by ConercaoAlpha** [Deletescape-Media/Lawnchair#3517]
### Describe the feature

Second screen on the right becomes part of the main screen.

On the pixel fold version of the Pixel Launcher, when the Fold is opened (if the user has more than one home screen) the additional screen becomes the right hand side of the larger unfolded screen instead of expanding the first home screen to fill the larger screen

### Why would it be useful to add?

This would be great to replicate a useful function of the Pixel Launcher 

### Did the feature exist in Lawnchair v2? (Play Store version)

No

### Example(s)

![image](https://github.com/LawnchairLauncher/lawnchair/assets/37657888/4779299d-8fca-4d14-9399-0bb210aab046)

Please note the dock is the same on both screen 1 and screen 2

**Comment by kbexit** [comment]
This is a need.
Current user of the Pixel Fold.

**Comment by rodrigozangelmi** [comment]
Would be nice!

**Comment by iamjonmiller** [comment]
I have a foldable (OnePlus Open) and access to development tools so I am going to take a crack at this.

**Comment by iamjonmiller** [comment]
By hardcoding the return of `getPanelCount()` to `2` I can confirm that there actually is working code within Lawnchair to merge 2 "screens" into 1 on Foldables.

![Screenshot 2024-09-08 134818](https://github.com/user-attachments/assets/84c4c8b3-c3da-4223-a3c5-f9c67579adaf)
![image](https://github.com/user-attachments/assets/7bf299e7-ac66-4846-9f37-7640191edf95)


What seems to be not working is the _detecting_ of a foldable. I am going to keep pulling on this string as it seems promising.

Basically, `isMultiDisplay` seems to always be `false` and thus it falls back to `isTablet` because of screen dimensions.

**Comment by iamjonmiller** [comment]
Fundamentally this traces back to `getDeviceType` in `InvariantDeviceProfile.java`

```
        public static @DeviceType int getDeviceType(Info displayInfo) {
                int flagPhone = 1 << 0;
                int flagTablet = 1 << 1;

                int type = displayInfo.supportedBounds.stream()
                        .mapToInt(bounds -> displayInfo.isTablet(bounds) ? flagTablet : flagPhone)
                        .reduce(0, (a, b) -> a | b);
                if ((type == (flagPhone | flagTablet))) {
                        // device has profiles supporting both phone and table modes
                        return TYPE_MULTI_DISPLAY;
                } else if (type == flagTablet) {
                        return TYPE_TABLET;
                } else {
                        return TYPE_PHONE;
                }
        }
```

While there is a loop that theoretically could return `TYPE_MULTI_DISPLAY` this never actually happens because `displayInfo.isTablet` only is aware of a single display at a time.

There is an unused function `getAllDisplays()` right next to `isTablet()` in `DisplayController.java` but it always returns empty.

If I can figure out how to actually look at all the available displays on a device (from my crude understanding `CachedDisplayInfo` seems to only store a single display which is overwritten when dimensions change) it seems like it would be pretty easy to rewrite the `getDeviceType` loop to take all displays into account. 

I just don't know enough about native Android dev to start writing a function to actually lookup all the device's displays, so I'm stuck here for now.

Now there is one other solution... just replace `TYPE_TABLET` with `TYPE_MULTI_DISPLAY`

```
        public static @DeviceType int getDeviceType(Info displayInfo) {
                int flagPhone = 1;
                int flagTablet = 1 << 1;

                int type = displayInfo.supportedBounds.stream()
                        .mapToInt(bounds -> displayInfo.isTablet(bounds) ? flagTablet : flagPhone)
                        .reduce(0, (a, b) -> a | b);
            
//                if ((type == (flagPhone | flagTablet))) {
//                        // device has profiles supporting both phone and table modes
//                        return TYPE_MULTI_DISPLAY;
                if (type == flagTablet) {
                        return TYPE_MULTI_DISPLAY;
                } else {
                        return TYPE_PHONE;
                }
        }
```

This just works. Perfect merging of pages with clean spacing.
![image](https://github.com/user-attachments/assets/ea2ba612-6ea8-4a78-af29-aa05577810d7)

Perfect transition to `TYPE_PHONE` once screen size changes.
![image](https://github.com/user-attachments/assets/c2bb30d4-d42a-4163-94c8-4c69439b5c1c)

And nothing is visibly broken on actual tablets, they just default to a 2 page homescreen which only really prevents centered widgets.
![image](https://github.com/user-attachments/assets/14ed31b9-2afe-4a13-a7f2-f8df57bb3ee6)

**Comment by iamjonmiller** [comment]
I have spent an entire evening reading through the `systemUnFold` library and trying to figure out a way to integrate the folded status into `DisplayController.java`. It's simply beyond me presently. I know what I want to do, I just can't figure out how to access and store the data in a way that it is accessible within other functions (within the `public static class Info` which is used in things like `getDeviceType`).

We just need a simple method to measure whether or not folding is applicable to the current display. It could be the hinge angle, the folding status (available in a variety of forms) or some method that defines a "default" display (either display seems to always show up as `0` or default presently).

**Comment by SuperDragonXD** [comment]
Something else you could do is to create a singleton class using the Jetpack library within the `lawnchair` module (example: FoldableCompat) and usinh that class within `InvariantDeviceProfile` (or similar) to get the details.

**Comment by vincentcomfy** [comment]
@iamjonmiller I'm sorry to poke you, but did you ever get anywhere with this? :( Really hope Lawnchair gets better fold support in the future.

**Comment by iamjonmiller** [comment]
> [@iamjonmiller](https://github.com/iamjonmiller) I'm sorry to poke you, but did you ever get anywhere with this? :( Really hope Lawnchair gets better fold support in the future.

I didn't come up with a real solution because this is beyond my Android dev abilities and it didn't seem like there was any real interest with the devs to make this a priority. I just changed the one variable I pointed out and compiled by own debug build which I use.

**Comment by Cerebus628** [comment]
Ressurecting this issue.  Any progress? It's properly implemented in Launcher3 aosp.   This is the only feature preventing me from using Lawnchair on my GrapheneOS install.  Attached are the photos of how launcher3 functions on a fold/tablet device.  

![Image](https://github.com/user-attachments/assets/85e33d65-7545-48d0-bedf-bfda9dfd8382)

![Image](https://github.com/user-attachments/assets/fb3f3f26-4937-499f-a891-eff5af5f0db9)

As you can see when the phone is unfolded the dock is extended to add additional apps and two screens are shown at once.  

![Image](https://github.com/user-attachments/assets/5d4c605e-db22-432f-bcc1-e47f93a0eb55)

Also, the toolbar is missing when using Lawnchair.

**Comment by Cerebus628** [comment]
The code is available in the aosp launcher3 which is why this works on GrapheneOS too.   

https://android.googlesource.com/platform/packages/apps/Launcher3/

https://github.com/GrapheneOS/platform_packages_apps_Launcher3

I bet the solution is somewhere in these codebases.  If not AOSP then definitely GrapheneOS.

**Comment by Zoomll** [comment]
This feature would be great! At the moment lawnchair is usable on a fold but it isn't very good looking. Hope this can be resolved soon

**Comment by DHernan** [comment]
Is this planned for 16 maybe since it's already in aosp?
